server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -417,89 +466,91 @@ private void loadAllLookupsAndInitStateRef()
         "LookupReferencesManager-Startup-%s"
     );
-    ExecutorCompletionService completionService = new ExecutorCompletionService(executorService);
-    List<Future<Map.Entry>> futures = new ArrayList<>();
+    CompletionService<Map.Entry<String, LookupExtractorFactoryContainer>> completionService =
+        new ExecutorCompletionService<>(executorService);
     try {
       LOG.info("Starting lookup loading process");
-      for (LookupBean lookupBean : lookupBeanList) {
-        futures.add(
-            completionService.submit(
-                () -> {
-                  LookupExtractorFactoryContainer container = lookupBean.getContainer();
-                  LOG.info(
-                      "Starting lookup [%s]:[%s]",
-                      lookupBean.getName(),
-                      container
-                  );
-                  if (container.getLookupExtractorFactory().start()) {
-                    LOG.info(
-                        "Started lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
-                  } else {
-                    LOG.error(
-                        "Failed to start lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return null;
-                  }
-                }
-            )
-        );
-      }
-      for (int i = 0; i < futures.size(); i++) {
-        try {
-          final Future<AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer>> completedFuture = completionService
-              .take();
-          final AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer> lookupResult = completedFuture
-              .get();
-          if (lookupResult != null) {
-            builder.put(lookupResult);
-          }
-        }
-        catch (ExecutionException e) {
-          LOG.error(e, "Execution error during lookup loading.");
+      List<LookupBean> remainingLookups = lookupBeanList;
+      for (int i = 0; i < lookupConfig.getLookupStartRetries(); i++) {
+        LOG.info("Round of attempts #%d, [%d] lookups", i + 1, remainingLookups.size());
+        Map<String, LookupExtractorFactoryContainer> successfulLookups =
+            startLookups(remainingLookups, completionService);
+        builder.putAll(successfulLookups);
+        List<LookupBean> failedLookups = remainingLookups
+            .stream()
+            .filter(l -> !successfulLookups.containsKey(l.getName()))
+            .collect(Collectors.toList());
+        if (failedLookups.isEmpty()) {
+          break;
+        } else {
+          // next round
+          remainingLookups = failedLookups;
         }
       }
       stateRef.set(new LookupUpdateState(builder.build(), ImmutableList.of(), ImmutableList.of()));
     }
```

Comment:
in the end can we log the failed lookups that we were not able to load even after all retries?

Reply:
Added